### ISSUE TITLE
Error: rpc error: code = Unavailable desc = transport is closing (interface conversion panic)

### ISSUE THREAD

**Issue by claudiobizzotto** [terraform-providers/terraform-provider-aws#11420]
### Terraform Version

```
Terraform v0.12.18
+ provider.aws v2.43.0
+ provider.template v2.1.2
```

### Affected Resource(s)

`s3_bucket` defined in a custom Terraform module.

### Terraform Configuration Files

Can't share template, but I can say it makes use of a Terraform module that defines an S3 bucket resource.

### Debug Output

```
Error: rpc error: code = Unavailable desc = transport is closing
```

### Panic Output

```
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: panic: interface conversion: interface {} is nil, not map[string]interface {}
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: goroutine 64 [running]:
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/terraform-providers/terraform-provider-aws/aws.expirationHash(0x0, 0x0, 0x1)
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_s3_bucket.go:2312 +0x3d6
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Set).hash(0xc000b5a5e0, 0x0, 0x0, 0x746172697078652e, 0x2e302e6e6f69)
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/set.go:221 +0x3d
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Set).add(0xc000b5a5e0, 0x0, 0x0, 0xc0006b3700, 0x0, 0xc000b19080)
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/set.go:198 +0x83
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*ConfigFieldReader).readSet(0xc000b4ef60, 0xc000b4f950, 0x3, 0x3, 0xc0006b37c0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/field_reader_config.go:309 +0x329
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*ConfigFieldReader).readField(0xc000b4ef60, 0xc000b4f950, 0x3, 0x3, 0xc000088801, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/field_reader_config.go:132 +0x97f
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*nestedConfigFieldReader).ReadField(0xc00000fac0, 0xc000b4f950, 0x3, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/field_reader_config.go:352 +0x8e
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.readObjectField(0x66d9b20, 0xc00000fac0, 0xc000b5a4e0, 0x2, 0x2, 0xc00068d110, 0x0, 0x0, 0x0, 0x0, ...)
2019-12-24T12:24:05.785-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/field_reader.go:244 +0x1fc
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*ConfigFieldReader).readField(0xc000b4ef60, 0xc000b5a4e0, 0x2, 0x2, 0xc000088801, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/field_reader_config.go:134 +0x8c7
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*nestedConfigFieldReader).ReadField(0xc00000fab0, 0xc000b5a4e0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/field_reader_config.go:352 +0x8e
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.readListField(0x66d9b20, 0xc00000fab0, 0xc000b54ec0, 0x1, 0x1, 0xc0006b4500, 0x0, 0x0, 0x0, 0x0, ...)
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/field_reader.go:211 +0x340
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*ConfigFieldReader).readField(0xc000b4ef60, 0xc000b54ec0, 0x1, 0x1, 0xc000b5a400, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/field_reader_config.go:128 +0x60f
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*ConfigFieldReader).ReadField(0xc000b4ef60, 0xc000b54ec0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x95005d8, ...)
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/field_reader_config.go:28 +0xae
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*MultiLevelFieldReader).ReadFieldExact(0xc00091fcc0, 0xc000b54ec0, 0x1, 0x1, 0x5e734c6, 0x6, 0x0, 0x0, 0x0, 0x0, ...)
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/field_reader_multi.go:31 +0xd0
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*ResourceData).get(0xc0001502a0, 0xc000b54ec0, 0x1, 0x1, 0x12, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/resource_data.go:534 +0xfa
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*ResourceData).getChange(0xc0001502a0, 0x5e8a03b, 0xe, 0xc000921201, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/resource_data.go:510 +0x122
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*ResourceData).diffChange(0xc0001502a0, 0x5e8a03b, 0xe, 0xc0006b2640, 0xc0009273e8, 0x67885c0, 0xc0001502a0, 0x1000000)
2019-12-24T12:24:05.786-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/resource_data.go:487 +0x97
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.schemaMap.diffList(0xc00068cff0, 0x5e8a03b, 0xe, 0xc0006b4500, 0xc0009273e8, 0x67885c0, 0xc0001502a0, 0x102cc00, 0x0, 0x0)
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/schema.go:958 +0x66
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.schemaMap.diff(0xc00068cff0, 0x5e8a03b, 0xe, 0xc0006b4500, 0xc00091fbc0, 0x67885c0, 0xc0001502a0, 0x6718100, 0x0, 0x0)
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/schema.go:919 +0x464
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.schemaMap.Diff(0xc00068cff0, 0xc000910460, 0xc000b4e420, 0x0, 0x4d195e0, 0xc0002e4500, 0x6773600, 0xc0008a6340, 0xffffffffffffffff, 0x0)
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/schema.go:500 +0x215
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).simpleDiff(0xc0006a5b80, 0xc000910460, 0xc000b4e420, 0x4d195e0, 0xc0002e4500, 0xc000b4e401, 0xc000927790, 0x100d0ed)
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/resource.go:351 +0x85
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).SimpleDiff(0xc0007af000, 0xc000927978, 0xc000910460, 0xc000b4e420, 0xc00090c1a0, 0xc000b4e420, 0x0)
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/helper/schema/provider.go:321 +0x99
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).PlanResourceChange(0xc00000f340, 0x6772a80, 0xc0009a2930, 0xc000aca720, 0xc00000f340, 0xc0009a2930, 0xc000851a80)
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/internal/helper/plugin/grpc_provider.go:633 +0x765
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_PlanResourceChange_Handler(0x5d3b160, 0xc00000f340, 0x6772a80, 0xc0009a2930, 0xc000aca6c0, 0x0, 0x6772a80, 0xc0009a2930, 0xc000b10e00, 0x6c6)
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.4.1/internal/tfplugin5/tfplugin5.pb.go:3171 +0x217
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: google.golang.org/grpc.(*Server).processUnaryRPC(0xc0003c2000, 0x6793e40, 0xc000550780, 0xc0004e8100, 0xc0007ad170, 0x94d44e8, 0x0, 0x0, 0x0)
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:995 +0x460
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: google.golang.org/grpc.(*Server).handleStream(0xc0003c2000, 0x6793e40, 0xc000550780, 0xc0004e8100, 0x0)
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:1275 +0xd97
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0001206d0, 0xc0003c2000, 0x6793e40, 0xc000550780, 0xc0004e8100)
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:710 +0xbb
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: created by google.golang.org/grpc.(*Server).serveStreams.func1
2019-12-24T12:24:05.787-0500 [DEBUG] plugin.terraform-provider-aws_v2.43.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:708 +0xa1
```

### Steps to Reproduce

`terraform apply`

### References

* #10084

**Comment by claudiobizzotto** [comment]
Looks like the error was in the way I was defaulting the bucket expiration. My shared module was configured with

```
expiration {
  days = var.days_before_expiration
}
```

And my `vars.tf` had this:

```
variable "days_before_expiration" {
  description = "Days before objects are permanently deleted."
  type        = number
  default     = null
}
```

Why `null` and not `0`? Because I was _hoping_ that, by passing that variable as `null`, the final resource would ignore the `expiration` block entirely, giving me the option of not setting the expiration for some buckets.

It didn't work, so now I'm using this:

```
variable "days_before_expiration" {
  description = "Days before objects are permanently deleted. Only effective when lifecycle management is enabled."
  type        = number
  default     = 36500 # 100 years
}
```

That works ;)

It would be nice to have a more obvious error message, though, so I'll leave the issue open.

**Comment by bflad** [comment]
The cause for the panic within the code is not verifying the passed in object is the expected type before performing the type assertion here:

https://github.com/terraform-providers/terraform-provider-aws/blob/b6d453f73fc978ce900198f13df646b13e4717ac/aws/resource_aws_s3_bucket.go#L2312

We likely want to be defensive for this and the other hash functions in this resource with the same issue by using the second return value for Go type assertions and exiting early if it does not match, e.g.

```go
m, ok := v.(map[string]interface{})
if !ok {
  return 0
}
```

This will only prevent the panic here though -- further handling of the empty configuration block may cause another panic or Terraform displaying a permanent difference in the existence of the configuration block (e.g. `.expiration.# 0 => 1`).

**Comment by lyle-partsavatar** [comment]
I seem to be having a similar problem.  I'm trying to create a Cloudfront configuration.  I'm using IntelliJ which provides autocomplete for required HCL attributes, and my code looks like this:

`resource "aws_cloudfront_distribution" "cloudfront_distribution" {
  enabled = true
  default_cache_behavior {
    allowed_methods = ["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
    cached_methods = ["GET", "HEAD"]
    target_origin_id = "*"
    viewer_protocol_policy = "redirect-to-https"
    forwarded_values {
      query_string = false
      cookies {
        forward = "all"
      }
    }
  }
  origin {
    domain_name = var.alb_domain_name
    origin_id = "<ALB_NAME>"
  viewer_certificate {}

    custom_header {
      name = "<HEADER_NAME>"
      value = <HEADER_VAL>
    }
  }
  restrictions {
    geo_restriction {
      restriction_type = ""
    }
  }
}`

I'm getting the above error when I try to `terraform apply`.  If I remove that code then everything works.

**Comment by timrourke** [comment]
@lyle-partsavatar I ran into the same issue, in my case I was able to get things working again by changing this:

```hcl
  restrictions {
    geo_restriction {
      restriction_type = ""
    }
  }
```

...to this:

```hcl
  restrictions {
    geo_restriction {
      restriction_type = "none"
    }
  }
```

**Comment by bflad** [comment]
Crash fix for original report with `aws_s3_bucket` resource submitted here: https://github.com/terraform-providers/terraform-provider-aws/pull/12614

The `aws_cloudfront_distribution` resource panic is likely related, but separate code, and can be tracked in this issue: https://github.com/terraform-providers/terraform-provider-aws/issues/10668

**Comment by bflad** [comment]
The fix mentioned above has been merged and will release with version 2.56.0 of the Terraform AWS Provider, later this week. 👍
